File: controllers/tagController.js
const Tag = require("../model/tag");
const AuditLog = require('../model/Auditlogs'); // Import the Audit Log model

// Create a new tag
exports.createTag = async (req, res) => {
  try {
    const tag = new Tag(req.body);
    await tag.save();

    // Create an audit log entry for the tag creation
    await AuditLog.create({
      userId: req.user._id,
      action: 'create',
      entity: 'Tag',
      entityId: tag._id,
      newValue: tag,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.status(201).json(tag);
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
};

// Get all tags
exports.getAllTagsByType = async (req, res) => {
  try {
    const tags = (await Tag.find({
      tagType: req.params.type,
    })).map((tag) => {
      return {
        _id: tag._id,
        label: tag.name,
      };
    });

    // Create an audit log entry for retrieving tags
    await AuditLog.create({
      userId: req.user._id,
      action: 'view',
      entity: 'Tag',
      entityId: null, // No specific entity ID for this action
      newValue: tags,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.status(200).json({ tags });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

// Get a specific tag
exports.getTagById = async (req, res) => {
  try {
    const tag = await Tag.findById(req.params.id);
    if (!tag) {
      return res.status(404).json({ error: "Tag not found" });
    }

    // Create an audit log entry for retrieving a specific tag
    await AuditLog.create({
      userId: req.user._id,
      action: 'view',
      entity: 'Tag',
      entityId: tag._id,
      newValue: tag,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.json(tag);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

// Update a tag
exports.updateTag = async (req, res) => {
  try {
    const tag = await Tag.findByIdAndUpdate(req.params.id, req.body, {
      new: true,
    });
    if (!tag) {
      return res.status(404).json({ error: "Tag not found" });
    }

    // Create an audit log entry for the tag update
    await AuditLog.create({
      userId: req.user._id,
      action: 'update',
      entity: 'Tag',
      entityId: tag._id,
      oldValue: { ...tag._doc }, // Store old values
      newValue: tag,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.json(tag);
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
};

// Delete a tag
exports.deleteTag = async (req, res) => {
  try {
    const tag = await Tag.findByIdAndDelete(req.params.id);
    if (!tag) {
      return res.status(404).json({ error: "Tag not found" });
    }

    // Create an audit log entry for the tag deletion
    await AuditLog.create({
      userId: req.user._id,
      action: 'delete',
      entity: 'Tag',
      entityId: tag._id,
      oldValue: tag,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.json({ message: "Tag deleted successfully" });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

// Search for tags by name
exports.searchTagsByName = async (req, res) => {
  try {
    const { name, type } = req.params;
    const tags = await Tag.find({
      name: { $regex: new RegExp(name, "i") },
      type,
      created_by: req.user._id
    });

    // Create an audit log entry for the search action
    await AuditLog.create({
      userId: req.user._id,
      action: 'search',
      entity: 'Tag',
      entityId: null, // No specific entity ID for this action
      newValue: tags,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.json(tags);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

File: controllers/tag_controller.py
from fastapi import HTTPException, Request
from models import Tag, AuditLog  # Assuming Beanie models
from typing import Dict, Any, List

class TagController:
    @staticmethod
    async def create_tag(data: Dict[str, Any], request: Request, user_id: str) -> Dict[str, Any]:
        """Create a new tag and log the action."""
        try:
            tag = Tag(**data)
            await tag.insert()

            await AuditLog(
                user_id=user_id,
                action="create",
                entity="tag",
                entity_id=str(tag.id),
                new_value=tag.dict(),
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return tag.dict()
        except Exception as e:
            raise HTTPException(status_code=400, detail=str(e))

    @staticmethod
    async def get_all_tags_by_type(tag_type: str, request: Request, user_id: str) -> Dict[str, Any]:
        """Retrieve all tags by type and log the action."""
        try:
            tags = await Tag.find(Tag.tag_type == tag_type).to_list()
            formatted_tags = [{"_id": str(tag.id), "label": tag.name, "type": tag.tag_type} for tag in tags]

            await AuditLog(
                user_id=user_id,
                action="view",
                entity="tag",
                entity_id=None,
                new_value=formatted_tags,
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return {"tags": formatted_tags}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @staticmethod
    async def get_tag_by_id(tag_id: str, request: Request, user_id: str) -> Dict[str, Any]:
        """Retrieve a specific tag and log the action."""
        try:
            tag = await Tag.get(tag_id)
            if not tag:
                raise HTTPException(status_code=404, detail="Tag not found")

            await AuditLog(
                user_id=user_id,
                action="view",
                entity="tag",
                entity_id=str(tag.id),
                new_value=tag.dict(),
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return tag.dict()
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @staticmethod
    async def update_tag(tag_id: str, data: Dict[str, Any], request: Request, user_id: str) -> Dict[str, Any]:
        """Update a tag and log the action."""
        try:
            tag = await Tag.get(tag_id)
            if not tag:
                raise HTTPException(status_code=404, detail="Tag not found")

            old_value = tag.dict()
            for key, value in data.items():
                setattr(tag, key, value)
            await tag.save()

            await AuditLog(
                user_id=user_id,
                action="update",
                entity="tag",
                entity_id=str(tag.id),
                old_value=old_value,
                new_value=tag.dict(),
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return tag.dict()
        except Exception as e:
            raise HTTPException(status_code=400, detail=str(e))

    @staticmethod
    async def delete_tag(tag_id: str, request: Request, user_id: str) -> Dict[str, str]:
        """Delete a tag and log the action."""
        try:
            tag = await Tag.get(tag_id)
            if not tag:
                raise HTTPException(status_code=404, detail="Tag not found")

            old_value = tag.dict()
            await tag.delete()

            await AuditLog(
                user_id=user_id,
                action="delete",
                entity="tag",
                entity_id=tag_id,
                old_value=old_value,
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return {"message": "Tag deleted successfully"}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @staticmethod
    async def search_tags_by_name(tag_type: str, name: str) -> List[Dict[str, Any]]:
        """Search tags by name."""
        try:
            tags = await Tag.find(
                Tag.name.regex(f"(?i){name}"),
                Tag.tag_type == tag_type
            ).to_list()
            return [tag.dict() for tag in tags]
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))